File: apps/mobile/lib/services/coach/coach_context_builder.dart
/// Coach Context Builder — Sprint S35 (Coach Narrative).
///
/// Builds a [CoachContext] from profile data for use by
/// [FallbackTemplates], [CoachNarrativeService], and [ComplianceGuard].
///
/// ## Numeric Grounding Contract
///
/// All values in [knownValues] MUST come from financial_core calculators
/// (never raw user input) to ensure [HallucinationDetector] can verify
/// LLM output against ground truth.
///
/// ### Grounding keys (registered in [knownValues]):
///
/// | Key                  | Source                       | Unit    | Tolerance |
/// |----------------------|------------------------------|---------|-----------|
/// | `fri_total`          | FinancialFitnessService      | 0-100   | ±2 pts   |
/// | `replacement_ratio`  | ForecasterService            | %       | ±2 pts   |
/// | `months_liquidity`   | derived (liquid / expenses)  | months  | ±5%      |
/// | `tax_saving`         | TaxCalculator                | CHF     | ±5%      |
/// | `confidence_score`   | ConfidenceScorer             | 0-100   | ±2 pts   |
/// | `capital_final`      | ForecasterService            | CHF     | ±5%      |
/// | `epargne_3a`         | Profile (verified)           | CHF     | ±5%      |
/// | `avoir_lpp`          | Profile (verified)           | CHF     | ±5%      |
/// | `salaire_brut`       | Profile (verified)           | CHF/an  | ±5%      |
///
/// Adding a value here means [HallucinationDetector] will flag any LLM
/// output containing a number that deviates beyond tolerance from these
/// ground-truth values. Only add values that are **computed or verified**
/// — never raw user text input.
///
/// References:
///   - FINMA circular 2008/21 (operational risk)
///   - LSFin art. 8 (quality of financial information)
library;

import 'coach_models.dart';

class CoachContextBuilder {
  CoachContextBuilder._();

  /// Build a [CoachContext] from profile-derived values.
  ///
  /// All numeric values should originate from financial_core calculators
  /// or derived computations — never from raw user text input.
  ///
  /// Parameters:
  ///   - [firstName]: User's first name (default: 'utilisateur')
  ///   - [age]: User's age
  ///   - [canton]: Swiss canton code (e.g. 'VD', 'ZH')
  ///   - [archetype]: Financial archetype (see ADR-20260223)
  ///   - [friTotal]: Current Financial Resilience Index score (0-100)
  ///   - [friDelta]: Change in FRI since last check-in
  ///   - [primaryFocus]: Current coaching focus area
  ///   - [replacementRatio]: Estimated retirement replacement ratio (0.0-1.0)
  ///   - [monthsLiquidity]: Months of expenses covered by liquid savings
  ///   - [taxSavingPotential]: Estimated tax savings from 3a/LPP actions (CHF)
  ///   - [confidenceScore]: Projection confidence (0-100)
  ///   - [capitalFinal]: Projected retirement capital (CHF)
  ///   - [epargne3a]: Current 3a savings balance (CHF)
  ///   - [avoirLpp]: Current LPP balance (CHF)
  ///   - [salaireBrut]: Annual gross salary (CHF)
  ///   - [daysSinceLastVisit]: Days since last app visit
  ///   - [fiscalSeason]: Current fiscal season code
  ///   - [checkInStreak]: Consecutive months of check-ins
  ///   - [lastMilestone]: Last achieved milestone label
  /// [dataSources]: Map from profile field key to source enum name
  ///   (e.g. {'prevoyance.avoirLppTotal': 'certificate'}).
  ///   Passed through as simplified dataReliability map for SLM prompting.
  static CoachContext build({
    String firstName = 'utilisateur',
    int age = 30,
    String canton = 'VD',
    String archetype = 'swiss_native',
    double friTotal = 0,
    double friDelta = 0,
    String primaryFocus = '',
    double replacementRatio = 0,
    double monthsLiquidity = 0,
    double taxSavingPotential = 0,
    double confidenceScore = 0,
    double capitalFinal = 0,
    double epargne3a = 0,
    double avoirLpp = 0,
    double salaireBrut = 0,
    int daysSinceLastVisit = 0,
    String fiscalSeason = '',
    int checkInStreak = 0,
    String lastMilestone = '',
    String upcomingEvent = '',
    Map<String, String> dataSources = const {},
  }) {
    // Build grounding values map — only include non-zero values
    // so HallucinationDetector doesn't false-positive on missing data.
    final knownValues = <String, double>{};
    if (friTotal > 0) knownValues['fri_total'] = friTotal;
    if (replacementRatio > 0) {
      knownValues['replacement_ratio'] = replacementRatio * 100;
    }
    if (monthsLiquidity > 0) knownValues['months_liquidity'] = monthsLiquidity;
    if (taxSavingPotential > 0) knownValues['tax_saving'] = taxSavingPotential;
    if (confidenceScore > 0) knownValues['confidence_score'] = confidenceScore;
    if (capitalFinal > 0) knownValues['capital_final'] = capitalFinal;
    if (epargne3a > 0) knownValues['epargne_3a'] = epargne3a;
    if (avoirLpp > 0) knownValues['avoir_lpp'] = avoirLpp;
    if (salaireBrut > 0) knownValues['salaire_brut'] = salaireBrut;

    return CoachContext(
      firstName: firstName,
      archetype: archetype,
      age: age,
      canton: canton,
      friTotal: friTotal,
      friDelta: friDelta,
      primaryFocus: primaryFocus,
      replacementRatio: replacementRatio,
      monthsLiquidity: monthsLiquidity,
      taxSavingPotential: taxSavingPotential,
      confidenceScore: confidenceScore,
      daysSinceLastVisit: daysSinceLastVisit,
      fiscalSeason: fiscalSeason,
      upcomingEvent: upcomingEvent,
      checkInStreak: checkInStreak,
      lastMilestone: lastMilestone,
      knownValues: knownValues,
      dataReliability: dataSources,
    );
  }
}

File: services/backend/app/services/coach/coach_context_builder.py
"""
Coach Context Builder — Sprint S35 (Coach Narrative Service).

Pure function that builds a CoachContext from profile data,
populating known_values for hallucination detection.

This is the ONLY entry point for creating CoachContext from user data.
Consumers (endpoints, services) call build_coach_context() and pass
the result to CoachNarrativeService.

Sources:
    - LSFin art. 3 (information financiere)
    - LPD art. 6 (protection des donnees)
"""

from typing import List, Optional

from app.services.coach.coach_models import CoachContext


def build_coach_context(
    first_name: str = "utilisateur",
    age: int = 30,
    canton: str = "VD",
    archetype: str = "swiss_native",
    fri_total: float = 0.0,
    fri_delta: float = 0.0,
    primary_focus: str = "",
    replacement_ratio: float = 0.0,
    months_liquidity: float = 0.0,
    tax_saving_potential: float = 0.0,
    confidence_score: float = 0.0,
    days_since_last_visit: int = 0,
    fiscal_season: str = "",
    upcoming_event: str = "",
    check_in_streak: int = 0,
    last_milestone: str = "",
    planned_contributions: Optional[List[dict]] = None,
) -> CoachContext:
    """Build CoachContext with known_values populated from financial indicators.

    The known_values dict is used by HallucinationDetector to verify
    that numbers in LLM output match financial_core computations.

    Args:
        first_name: User's first name (default: "utilisateur").
        age: User's age.
        canton: Canton of fiscal residence.
        archetype: Financial archetype (swiss_native, expat_eu, etc.).
        fri_total: FRI total score (0-100).
        fri_delta: FRI change since last check-in.
        primary_focus: Current financial priority.
        replacement_ratio: Estimated retirement replacement ratio (0-1).
        months_liquidity: Months of liquidity reserve.
        tax_saving_potential: Potential tax saving from 3a (CHF).
        confidence_score: Projection confidence score (0-100).
        days_since_last_visit: Days since last app visit.
        fiscal_season: Current fiscal season ("3a_deadline", "tax_declaration", "").
        upcoming_event: Upcoming life event (if any).
        check_in_streak: Consecutive check-in count.
        last_milestone: Last achieved milestone.
        planned_contributions: List of planned monthly contributions (id, label, amount, category).

    Returns:
        CoachContext with all fields and known_values populated.
    """
    known = {
        "fri_total": fri_total,
        "replacement_ratio": replacement_ratio * 100,  # stored as percentage
        "months_liquidity": months_liquidity,
        "tax_saving": tax_saving_potential,
        "confidence_score": confidence_score,
    }

    return CoachContext(
        first_name=first_name,
        age=age,
        canton=canton,
        archetype=archetype,
        fri_total=fri_total,
        fri_delta=fri_delta,
        primary_focus=primary_focus,
        replacement_ratio=replacement_ratio,
        months_liquidity=months_liquidity,
        tax_saving_potential=tax_saving_potential,
        confidence_score=confidence_score,
        days_since_last_visit=days_since_last_visit,
        fiscal_season=fiscal_season,
        upcoming_event=upcoming_event,
        check_in_streak=check_in_streak,
        last_milestone=last_milestone,
        planned_contributions=planned_contributions or [],
        known_values=known,
    )
